avatar: Make system bot avatar URLs follow convention.

PieterCK opened the pull request:
This PR primarily makes requests to `get_avatar_url` for system bots avatars to behave more conventionaly.
- all system bots avatar files are now PNGs and have their "*-medium.png" variants
- requests with `medium=True` will return avatar URL with "*-medium.png" with image file size 500x500px. 
e.g `/static/images/static_avatars/notification-bot.aae0d3673451-medium.png`

Fixes: [CZO](https://chat.zulip.org/#narrow/channel/48-mobile/topic/notification.20bot.20icon)



**Screenshots and screen captures:**
<details>

<summary>new system bot avatar images specs</summary>

`-medium` files are 500x500px while the default ones are `100x100px`

```bash
identify ./static/images/static_avatars/*
./static/images/static_avatars/emailgateway-medium.png PNG 500x500 500x500+0+0 8-bit sRGB 160025B 0.000u 0:00.000
./static/images/static_avatars/emailgateway.png PNG 100x100 100x100+0+0 8-bit sRGB 5651B 0.000u 0:00.000
./static/images/static_avatars/notification-bot-medium.png PNG 500x500 500x500+0+0 16-bit sRGB 28115B 0.000u 0:00.000
./static/images/static_avatars/notification-bot.png PNG 100x100 100x100+0+0 16-bit sRGB 4875B 0.000u 0:00.000
./static/images/static_avatars/welcome-bot-medium.png PNG 500x500 500x500+0+0 8-bit sRGB 46c 23074B 0.000u 0:00.000
./static/images/static_avatars/welcome-bot.png PNG 100x100 100x100+0+0 8-bit sRGB 9065B 0.000u 0:00.000
```
</details>

<details>
<summary>avatar image in realm</summary>

![image](https://github.com/user-attachments/assets/084efe87-8651-4a0b-82cb-776b147cf92d)
![image](https://github.com/user-attachments/assets/df965e3b-a8fd-40eb-929a-d2c8f92dd392)

![image](https://github.com/user-attachments/assets/3c64f5a0-dd80-4ee7-8090-ce998915dbf0)
![image](https://github.com/user-attachments/assets/c845acb4-08aa-4e88-a731-de52d9fa6ac2)

I couldn't find where emailgateway bot is, but I'm sure they have the same resolution because I had to resize its avatar images too.

The actual image files can be seen in the file changes
</details>

<details>
<summary>manual tests</summary>

Rendering "*-medium.png" variants:
![image](https://github.com/user-attachments/assets/16652abb-532d-445c-9967-fc2b0bbd1f21)

---
Calling `/avatar/{email}/medium` will return the avatar URL with "-medium.png"

[Screencast from 10-22-2024 10:13:03 PM.webm](https://github.com/user-attachments/assets/3ad72e08-c34e-400f-ae53-53915ef35be9)

---
We get the hashed version of static avatar files (DEBUG=off):

[Screencast from 10-22-2024 10:24:43 PM.webm](https://github.com/user-attachments/assets/065d013b-916c-43e2-b895-e7ab88b3d1dc)

</details>

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

PieterCK commented on static/images/email-gateway-bot.png:
git grep "images/email-gateway-bot.png" returns nothing

PieterCK commented on static/images/welcome-bot.png:
git grep "images/welcome-bot.png"
returns nothing

PieterCK commented on zerver/lib/avatar.py:
```diff
@@ -36,6 +34,33 @@ def avatar_url(
+        from django.contrib.staticfiles.finders import find
+
+        if not find(avatar_file_name):
+            raise AssertionError(f"Unknown avatar file: {avatar_file_name}")
```
This is just a net to catch problems early on if any breaking changes to this system. For example, the avatar file name got changed, they moved, the system bots user email got updated, etc.

PieterCK commented on zerver/lib/avatar.py:
```diff
@@ -36,6 +34,33 @@ def avatar_url(
+        # not defined.
+        return DEFAULT_AVATAR_FILE
+    else:
+        from django.contrib.staticfiles.finders import find
```
I couldn't import `get_staticfile` here due to circular import.

PieterCK commented on zerver/lib/avatar.py:
```diff
@@ -36,6 +34,33 @@ def avatar_url(
 
 
+def get_system_bots_avatar_file_name(email: str) -> str:
+    # System bot avatar files are stored in the `images/static_avatars/`
```
we previously used dict for this logic
```
SYSTEM_BOTS_AVATAR_FILES = {
    # This is also used in zerver/lib/storage.py to ensure
    # these files are hashed when served as static files.
    settings.WELCOME_BOT: "images/welcome-bot.png",
    settings.NOTIFICATION_BOT: "images/logo/zulip-icon-square.svg",
    settings.EMAIL_GATEWAY_BOT: "images/email-gateway-bot.png",
}
```
I thought it would be a bit cleaner if we just matched the email username name, I wonder if it's better to just stick to dict for simplicity..

PieterCK commented on zerver/lib/avatar.py:
```diff
@@ -36,6 +34,33 @@ def avatar_url(
+    else:
+        from django.contrib.staticfiles.finders import find
+
+        if not find(avatar_file_name):
```
I couldn't import `static_path` for this due to circular import.

PieterCK commented:
Some logic in `avatar.py` is organized such that we can easily extend it once we implement the new avatar_source for this type of case. Once we have the migration I imagine it'd look like something like this:
<details>

<summary>git diff</summary>

```diff
diff --git a/zerver/lib/avatar.py b/zerver/lib/avatar.py
index f0f519d274..c35e671442 100644
--- a/zerver/lib/avatar.py
+++ b/zerver/lib/avatar.py
@@ -44,7 +44,10 @@ def get_system_bots_avatar_file_name(email: str) -> str:
 
 
 def get_static_avatar_url(email: str, medium: bool) -> str:
-    avatar_file_name = get_system_bots_avatar_file_name(email)
+    if is_cross_realm_bot_email(email):
+        avatar_file_name = get_system_bots_avatar_file_name(email)
+
+    # We can add another protocol for users of this avatar source here
 
     avatar_file_name += "-medium.png" if medium else ".png"
 
@@ -88,7 +91,7 @@ def get_avatar_field(
     """
 
     # System bots have hardcoded avatars
-    if is_cross_realm_bot_email(email):
+    if avatar_source == UserProfile.AVATAR_FROM_DISK:
         return get_static_avatar_url(email, medium)
 
     """
```
</details>

PieterCK commented on zerver/lib/storage.py:
```diff
@@ -40,12 +57,16 @@ def hashed_name(
             # assets.
             return name
-        if name in SYSTEM_BOTS_AVATAR_FILES.values():
+        if name.startswith(STATIC_AVATARS_DIR):
```
I simplified the logic to hash these avatar URLs, if it's in the `static/images/static_avatars/` dir, it'd get hashed. I wonder if the naming for this folder makes sense, I didn't use disk_avatar or avatar_disk to avoid confusion with `avatar_disk_path()` function in testing..

timabbott replied:
Yeah that seems cleaner.

PieterCK commented on zerver/lib/avatar.py:
```diff
@@ -36,6 +34,33 @@ def avatar_url(
+    else:
+        from django.contrib.staticfiles.finders import find
+
+        if not find(avatar_file_name):
```
I couldn't import `static_path` for this due to circular import.

timabbott replied:
I don't think we want to call this `find` operation on every request for the relevant avatar; that function may be expensive. I think it would be reasonable to put this as `if DEBUG and not find(...)`? 

The other option would be to try to do it at most once per process, rather than once per request, but I think that'd be more complexity than it's worth.

PieterCK commented on zerver/lib/avatar.py:
```diff
@@ -36,6 +34,33 @@ def avatar_url(
+        from django.contrib.staticfiles.finders import find
+
+        if not find(avatar_file_name):
+            raise AssertionError(f"Unknown avatar file for: {email}")
```
This is just a net to catch problems early on if any breaking changes to how we handle system bots avatar. For example, the avatar file name got changed, they moved, etc.

PieterCK commented on static/images/email-gateway-bot.png:
git grep "images/email-gateway-bot.png" returns nothing

PieterCK commented on zerver/lib/avatar.py:
git grep "images/welcome-bot.png"
returns nothing

timabbott commented on zerver/lib/storage.py:
```diff
@@ -24,6 +24,23 @@ def static_path(path: str) -> str:
+    Unlike user uploaded avatar files, we have to reformat
+    medium static avatars files so that the "-medium" token
+    is at the end of the name and not the middle.
+    """
```
I think we can skip this; the API doesn't specify anything about the URL format of avatars.

gnprice replied:
Well, the API for getting an avatar of different sizes isn't documented. In the absence of docs, I think this "-medium.png" behavior currently is part of the effective API — it's what the mobile apps have implemented, based on reverse-engineering long ago.

See the [discussion last year](https://chat.zulip.org/#narrow/channel/412-api-documentation/topic/avatar_url.3A.20Request.20specific.20size/near/1636546) that Chris [linked to](https://chat.zulip.org/#narrow/channel/48-mobile/topic/notification.20bot.20icon/near/1966399) in yesterday's thread.

As discussed in that previous thread, we could document something cleaner and then move the mobile apps to that. But I think in the meantime we'll need to support the existing de facto API for compatibility.

timabbott replied:
OK, I guess we can keep this commit.

timabbott replied:
Rewrote the comment to explain that context, since that'll be important for maintaining this properly.

timabbott commented:
I pushed a reworking of the logic in `avatar.py` in the first commit to use `settings.DEBUG` rather than reaching into `configured_settings.py`, and avoid calling `find` in production for every request to these system bot avatar URLs.

I also dropped and then re-added the second commit, see discussion above on why that is indeed necessary. It would have been nice for you to have referenced the prior conversation on it. I've rewritten the comments and commit message to explain why that logic is important and correct; what you had previously really didn't spell it out.

With those changes, this looks ready to merge.

timabbott commented:
Hmm, it looks like my attempt to clean up the `DEBUG` tooling breaks tests more than I'd expected. @PieterCK do you want to do the next pass on it? I think a possible theory is we should get rid of the static files check and assertion as not worth it.

PieterCK commented:
Changes:
- Checked `PRODUCTION` instead because it turns out Django always turns `DEBUG` off when running tests:
  >https://docs.djangoproject.com/en/5.1/topics/testing/overview/ :
  Regardless of the value of the DEBUG setting in your configuration file, all Django tests run with DEBUG =False. This is to ensure that the observed output of ...
  
  <details>
  
  <summary> git diff of the first commit </summary>
  
  ```diff
  diff --git a/zerver/lib/avatar.py b/zerver/lib/avatar.py
  index 99f2dbee15..ae2efe6db6 100644
  --- a/zerver/lib/avatar.py
  +++ b/zerver/lib/avatar.py
  @@ -46,18 +46,14 @@ def get_static_avatar_url(email: str, medium: bool) -> str:
       avatar_file_name = get_system_bots_avatar_file_name(email)
       avatar_file_name += "-medium.png" if medium else ".png"
   
  -    if settings.DEBUG:
  -        # This find call may not be cheap, so we only do it in the
  -        # development environment to do an assertion.
  -        from django.contrib.staticfiles.finders import find
  -
  -        if not find(avatar_file_name):
  -            raise AssertionError(f"Unknown avatar file for: {email}")
  -    elif not staticfiles_storage.exists(avatar_file_name):
  +    if settings.PRODUCTION and not staticfiles_storage.exists(avatar_file_name):
           # Fallback for the case where no avatar exists; this should
           # never happen in practice. This logic cannot be executed
  -        # while STATIC_ROOT is not defined, so the above DEBUG check
  -        # is important.
  +        # while STATIC_ROOT is not defined, so the above PRODUCTION
  +        # check is important.
  +        #
  +        # DEBUG isn't used here because it can be True when running
  +        # test even in development environment.
           return DEFAULT_AVATAR_FILE
   
       return staticfiles_storage.url(avatar_file_name)
  diff --git a/zerver/tests/test_upload.py b/zerver/tests/test_upload.py
  index f65d2faf90..616cef286a 100644
  --- a/zerver/tests/test_upload.py
  +++ b/zerver/tests/test_upload.py
  @@ -1557,13 +1557,8 @@ class AvatarTest(UploadSerializeMixin, ZulipTestCase):
               avatar_url_medium = get_static_avatar_url(email, True)
               self.assertNotIn(default_avatar, avatar_url_medium)
   
  -        with self.assertRaises(AssertionError) as e:
  -            avatar_url = get_static_avatar_url("false-bot@zulip.com", False)
  -        expected_error_message = "Unknown avatar file for: false-bot@zulip.com"
  -        self.assertEqual(str(e.exception), expected_error_message)
  -
           with (
  -            patch("zerver.lib.avatar.DEBUG", new=False),
  +            self.settings(PRODUCTION=True),
               patch("zerver.lib.avatar.staticfiles_storage.exists") as mock_exists,
           ):
               mock_exists.return_value = False
  ```
  </details>

#### - Extracted the assertion that checks for avatar files during dev to its own commit in [364ea0c](https://github.com/zulip/zulip/pull/32085/commits/364ea0c71b99eae5425a61f8f2dad1be8989bc07). Because we check for `PRODUCTION`, this doesn't introduce a lot of headaches when testing things--but it is safe to leave out.

#### Cleaned up the third commit, because we're checking `PRODUCTION` and not `DEBUG`, this simplifies things here, so we don't need to patch these anymore:
 <details>

<summary> git diff of the third commit </summary>

```diff
git diff 3975856 9e0315f -- zerver/tests/test_upload.py 
diff --git a/zerver/tests/test_upload.py b/zerver/tests/test_upload.py
index 87b69850a4..616cef286a 100644
--- a/zerver/tests/test_upload.py
+++ b/zerver/tests/test_upload.py
@@ -11,7 +11,6 @@ import pyvips
 import time_machine
 from django.conf import settings
 from django.core.files.uploadedfile import SimpleUploadedFile
-from django.test import override_settings
 from django.utils.timezone import now as timezone_now
 from pyvips import at_least_libvips
 from pyvips import version as libvips_version
@@ -1224,7 +1223,6 @@ class AvatarTest(UploadSerializeMixin, ZulipTestCase):
             redirect_url = response["Location"]
             self.assertEqual(redirect_url, "http://other.server/avatar.svg?version=1&foo=bar")
 
-    @override_settings(STATIC_ROOT="static/")
     def test_get_user_avatar(self) -> None:
         hamlet = self.example_user("hamlet")
         self.login_user(hamlet)
@@ -1539,7 +1537,6 @@ class AvatarTest(UploadSerializeMixin, ZulipTestCase):
             result = self.client_post("/json/users/me/avatar", {"file": fp})
         self.assert_json_error(result, "Uploaded file is larger than the allowed limit of 0 MiB")
 
-    @override_settings(STATIC_ROOT="static/")
     def test_system_bot_avatars_url(self) -> None:
         system_bot_emails = [
             settings.NOTIFICATION_BOT,
@@ -1560,12 +1557,8 @@ class AvatarTest(UploadSerializeMixin, ZulipTestCase):
             avatar_url_medium = get_static_avatar_url(email, True)
             self.assertNotIn(default_avatar, avatar_url_medium)
 
-        with self.settings(DEBUG=True), self.assertRaises(AssertionError) as e:
-            avatar_url = get_static_avatar_url("false-bot@zulip.com", False)
-        expected_error_message = "Unknown avatar file for: false-bot@zulip.com"
-        self.assertEqual(str(e.exception), expected_error_message)
-
         with (
+            self.settings(PRODUCTION=True),
             patch("zerver.lib.avatar.staticfiles_storage.exists") as mock_exists,
         ):
             mock_exists.return_value = False

```
</details>

PieterCK commented:
Ah, it turns out Django always turns `DEBUG` off when running tests:
  >https://docs.djangoproject.com/en/5.1/topics/testing/overview/ :
  Regardless of the value of the DEBUG setting in your configuration file, all Django tests run with DEBUG =False. This is to ensure that the observed output of ...
 
Updates:
- Extracted that checks for avatar files from the first commit into its own commit in [364ea0c](https://github.com/zulip/zulip/pull/32085/commits/364ea0c71b99eae5425a61f8f2dad1be8989bc07). 
 
<details>

<summary> git diff of the first commit </summary>

```diff

```
</details>

- In the second commit, I use `PRODUCTION` instead of `DEBUG` because of the reason I mentioned earlier. This ended up simplifying things because we don't need to `patch` everywhere. But, we can also leave this out as suggested. 

- Cleaned up the third commit :
   <details>
 
  <summary> git diff of the third commit </summary>
   git diff 3975856 00fddd0 -- zerver/tests/test_upload.py 
  diff --git a/zerver/tests/test_upload.py b/zerver/tests/test_upload.py
  index 87b69850a4..1800324bf5 100644
  --- a/zerver/tests/test_upload.py
  +++ b/zerver/tests/test_upload.py
  @@ -11,7 +11,6 @@ import pyvips
   import time_machine
   from django.conf import settings
   from django.core.files.uploadedfile import SimpleUploadedFile
  -from django.test import override_settings
   from django.utils.timezone import now as timezone_now
   from pyvips import at_least_libvips
   from pyvips import version as libvips_version
  @@ -1224,7 +1223,6 @@ class AvatarTest(UploadSerializeMixin, ZulipTestCase):
               redirect_url = response["Location"]
               self.assertEqual(redirect_url, "http://other.server/avatar.svg?version=1&foo=bar")
   
  -    @override_settings(STATIC_ROOT="static/")
       def test_get_user_avatar(self) -> None:
           hamlet = self.example_user("hamlet")
           self.login_user(hamlet)
  @@ -1539,7 +1537,6 @@ class AvatarTest(UploadSerializeMixin, ZulipTestCase):
               result = self.client_post("/json/users/me/avatar", {"file": fp})
           self.assert_json_error(result, "Uploaded file is larger than the allowed limit of 0 MiB")
   
  -    @override_settings(STATIC_ROOT="static/")
       def test_system_bot_avatars_url(self) -> None:
           system_bot_emails = [
               settings.NOTIFICATION_BOT,
  @@ -1560,18 +1557,19 @@ class AvatarTest(UploadSerializeMixin, ZulipTestCase):
               avatar_url_medium = get_static_avatar_url(email, True)
               self.assertNotIn(default_avatar, avatar_url_medium)
   
  -        with self.settings(DEBUG=True), self.assertRaises(AssertionError) as e:
  -            avatar_url = get_static_avatar_url("false-bot@zulip.com", False)
  -        expected_error_message = "Unknown avatar file for: false-bot@zulip.com"
  -        self.assertEqual(str(e.exception), expected_error_message)
  -
           with (
  +            self.settings(PRODUCTION=True),
               patch("zerver.lib.avatar.staticfiles_storage.exists") as mock_exists,
           ):
               mock_exists.return_value = False
               avatar_url = get_static_avatar_url("false-bot@zulip.com", False)
           self.assertIn(default_avatar, avatar_url)
   
  +        with self.assertRaises(AssertionError) as e:
  +            avatar_url = get_static_avatar_url("false-bot@zulip.com", False)
  +        expected_error_message = "Unknown avatar file for: false-bot@zulip.com"
  +        self.assertEqual(str(e.exception), expected_error_message)

  ```diff
  git diff 3975856 9e0315f -- zerver/tests/test_upload.py 
  diff --git a/zerver/tests/test_upload.py b/zerver/tests/test_upload.py
  index 87b69850a4..616cef286a 100644
  --- a/zerver/tests/test_upload.py
  +++ b/zerver/tests/test_upload.py
  @@ -11,7 +11,6 @@ import pyvips
   import time_machine
   from django.conf import settings
   from django.core.files.uploadedfile import SimpleUploadedFile
  -from django.test import override_settings
   from django.utils.timezone import now as timezone_now
   from pyvips import at_least_libvips
   from pyvips import version as libvips_version
  @@ -1224,7 +1223,6 @@ class AvatarTest(UploadSerializeMixin, ZulipTestCase):
               redirect_url = response["Location"]
               self.assertEqual(redirect_url, "http://other.server/avatar.svg?version=1&foo=bar")
   
  -    @override_settings(STATIC_ROOT="static/")
       def test_get_user_avatar(self) -> None:
           hamlet = self.example_user("hamlet")
           self.login_user(hamlet)
  @@ -1539,7 +1537,6 @@ class AvatarTest(UploadSerializeMixin, ZulipTestCase):
               result = self.client_post("/json/users/me/avatar", {"file": fp})
           self.assert_json_error(result, "Uploaded file is larger than the allowed limit of 0 MiB")
   
  -    @override_settings(STATIC_ROOT="static/")
       def test_system_bot_avatars_url(self) -> None:
           system_bot_emails = [
               settings.NOTIFICATION_BOT,
  @@ -1560,12 +1557,8 @@ class AvatarTest(UploadSerializeMixin, ZulipTestCase):
               avatar_url_medium = get_static_avatar_url(email, True)
               self.assertNotIn(default_avatar, avatar_url_medium)
   
  -        with self.settings(DEBUG=True), self.assertRaises(AssertionError) as e:
  -            avatar_url = get_static_avatar_url("false-bot@zulip.com", False)
  -        expected_error_message = "Unknown avatar file for: false-bot@zulip.com"
  -        self.assertEqual(str(e.exception), expected_error_message)
  -
           with (
  +            self.settings(PRODUCTION=True),
               patch("zerver.lib.avatar.staticfiles_storage.exists") as mock_exists,
           ):
               mock_exists.return_value = False
  
  ```
  </details>

PieterCK commented on zerver/tests/test_upload.py:
```diff
@@ -1540,23 +1545,33 @@ def test_system_bot_avatars_url(self) -> None:
             redirect_url = response["Location"]
             self.assertEqual(redirect_url, str(avatar_url(system_bot, medium=True)))
-            self.assertTrue(str(redirect_url).endswith("-medium.png"))
-            self.assertFalse(str(redirect_url).endswith("unknown-medium.png"))
```
these are deleted because the tests below will cover that. This is added just incase we don't want the second commit (add checks to make sure these files exists)

PieterCK commented:
Ah, it turns out Django turns `DEBUG` off when running tests:
  >https://docs.djangoproject.com/en/5.1/topics/testing/overview/ :
  Regardless of the value of the DEBUG setting in your configuration file, all Django tests run with DEBUG =False. This is to ensure that the observed output of ...
 
Updates:
- Extracted that checks for avatar files from the first commit into its own commit in [21e9d8d](https://github.com/zulip/zulip/pull/32085/commits/21e9d8d73cae2d118f44a060989750aab8454349)
 
  <details>

  <summary> git diff of the first commit </summary>
  
  ```diff
  git diff a27854e e9c16e0
  diff --git a/zerver/lib/avatar.py b/zerver/lib/avatar.py
  index 99f2dbee15..33533af0fc 100644
  --- a/zerver/lib/avatar.py
  +++ b/zerver/lib/avatar.py
  @@ -46,20 +46,6 @@ def get_static_avatar_url(email: str, medium: bool) -> str:
       avatar_file_name = get_system_bots_avatar_file_name(email)
       avatar_file_name += "-medium.png" if medium else ".png"
   
  -    if settings.DEBUG:
  -        # This find call may not be cheap, so we only do it in the
  -        # development environment to do an assertion.
  -        from django.contrib.staticfiles.finders import find
  -
  -        if not find(avatar_file_name):
  -            raise AssertionError(f"Unknown avatar file for: {email}")
  -    elif not staticfiles_storage.exists(avatar_file_name):
  -        # Fallback for the case where no avatar exists; this should
  -        # never happen in practice. This logic cannot be executed
  -        # while STATIC_ROOT is not defined, so the above DEBUG check
  -        # is important.
  -        return DEFAULT_AVATAR_FILE
  -
       return staticfiles_storage.url(avatar_file_name)
   
   
  diff --git a/zerver/tests/test_upload.py b/zerver/tests/test_upload.py
  index f65d2faf90..fdb241dbb9 100644
  --- a/zerver/tests/test_upload.py
  +++ b/zerver/tests/test_upload.py
  @@ -1557,19 +1557,6 @@ class AvatarTest(UploadSerializeMixin, ZulipTestCase):
               avatar_url_medium = get_static_avatar_url(email, True)
               self.assertNotIn(default_avatar, avatar_url_medium)
   
  -        with self.assertRaises(AssertionError) as e:
  -            avatar_url = get_static_avatar_url("false-bot@zulip.com", False)
  -        expected_error_message = "Unknown avatar file for: false-bot@zulip.com"
  -        self.assertEqual(str(e.exception), expected_error_message)
  -
  -        with (
  -            patch("zerver.lib.avatar.DEBUG", new=False),
  -            patch("zerver.lib.avatar.staticfiles_storage.exists") as mock_exists,
  -        ):
  -            mock_exists.return_value = False
  -            avatar_url = get_static_avatar_url("false-bot@zulip.com", False)
  -        self.assertIn(default_avatar, avatar_url)

  ```
  </details>

- In the second commit, I use `PRODUCTION` instead of `DEBUG` because of the reason I mentioned earlier. This ended up simplifying things because we don't need to `patch` everywhere. But, we can also leave this out as suggested. 

- Cleaned up the third commit :
   <details>
 
  <summary> git diff of the third commit </summary>

  ```diff
  git diff 3975856 1f41665 -- zerver/tests/test_upload.py
  diff --git a/zerver/tests/test_upload.py b/zerver/tests/test_upload.py
  index 87b69850a4..dd9b2229ec 100644
  --- a/zerver/tests/test_upload.py
  +++ b/zerver/tests/test_upload.py
  @@ -11,7 +11,6 @@ import pyvips
   import time_machine
   from django.conf import settings
   from django.core.files.uploadedfile import SimpleUploadedFile
  -from django.test import override_settings
   from django.utils.timezone import now as timezone_now
   from pyvips import at_least_libvips
   from pyvips import version as libvips_version
  @@ -1224,7 +1223,6 @@ class AvatarTest(UploadSerializeMixin, ZulipTestCase):
               redirect_url = response["Location"]
               self.assertEqual(redirect_url, "http://other.server/avatar.svg?version=1&foo=bar")
   
  -    @override_settings(STATIC_ROOT="static/")
       def test_get_user_avatar(self) -> None:
           hamlet = self.example_user("hamlet")
           self.login_user(hamlet)
  @@ -1539,38 +1537,40 @@ class AvatarTest(UploadSerializeMixin, ZulipTestCase):
               result = self.client_post("/json/users/me/avatar", {"file": fp})
           self.assert_json_error(result, "Uploaded file is larger than the allowed limit of 0 MiB")
   
  -    @override_settings(STATIC_ROOT="static/")
       def test_system_bot_avatars_url(self) -> None:

  ```
  </details>

PieterCK commented on zerver/lib/avatar.py:
```diff
@@ -36,6 +33,39 @@ def avatar_url(
+    avatar_file_name = get_system_bots_avatar_file_name(email)
+    avatar_file_name += "-medium.png" if medium else ".png"
+
+    if settings.PRODUCTION and not staticfiles_storage.exists(avatar_file_name):
```
If using `settings.PRODUCTION` is somehow problematic here I think checking for `settings.STATIC_ROOT` directly might also be another option,.

timabbott commented on zerver/tests/test_upload.py:
```diff
@@ -1532,6 +1532,31 @@ def test_avatar_upload_file_size_error(self) -> None:
+            self.assertTrue(str(redirect_url).endswith("-medium.png"))
+            self.assertFalse(str(redirect_url).endswith("unknown-medium.png"))
+
+        self.logout()
```
You don't need to logout at the end of a test class; the state is discarded via each test running inside a database transaction.
